### DISCUSSION TITLE
Implement Push API using Mozilla's servers

### DISCUSSION THREAD

**Discussion by notpushkin** [Eloston/ungoogled-chromium#1900] [Ideas]
In Chrome, Push API is implemented using FCM (a rebranding of GCM), which naturally doesn't work with Google's servers disabled. It might be an option to make them available again by using some third-party server, for example, the one used in Firefox ([Mozilla Push Service](https://mozilla-push-service.readthedocs.io/en/latest/#mozilla-push-service)).

Note that while it implements the same developer-facing API, the way it interacts with the browser might be a bit different. Depending on these differences, either patches to the browser code would be required, or a custom push server that re-implements FCM-compatible browser API might be a more feasible solution in some circumstances.

Previously: #589, #703.

**Comment by notpushkin** [comment]
> But does Chromium or a derivative implement the Push API? I'm not too familiar with it.

Basically, you can think of Push API as a vendor-neutral name for FCM/GCM (for websites at least, the `chrome.gcm` API in extensions is a completely different matter).

**Comment by lipici** [comment]
anything that has mozilla in name its a NO

**Comment by notpushkin** [comment]
@lipici Why not? (And in any case their implementation is open source, so we could run our own instance)

**Comment by Ph0rk0z** [comment]
This is stuff I turn off. No thanks on push notifications for the browser, especially on 3rd party servers.

**Comment by notpushkin** [comment]
@Ph0rk0z Usually I turn it off too, however it does make sense for some sites (messengers, notably), and some people have requested it (see #589, #703).

I agree with your privacy concerns though, so I'd make this an opt-in feature (i. e. make whatever option in `chrome://settings/content` allows pushes default to Disabled), this still shouldn't be too hard to implement.

**Comment by Ph0rk0z** [comment]
Does it make a big difference? I used messengers in FF and uGc and never had push notifications enabled.

**Comment by notpushkin** [comment]
@Ph0rk0z Well, not like you'll be obliged to use it :)

And if literally nobody needs it, nobody will work on it – I'm just pointing out the possibility, but I'm not _that_ interested myself, too (although I'd gladly work on a server part as that [might or might not be useful for microG too](https://github.com/microg/android_packages_apps_GmsCore/issues/934#issuecomment-622426399)).

**Comment by Eloston** [comment]
This is a good idea but I'm also not that interested in this...

Also, I think we should leave the default blank and let the user add the Mozilla Push API service if they want to.

**Comment by sevmonster** [comment]
Rather than use Mozilla servers, [roll your own](https://github.com/mozilla-services/autopush). I don't know if anyone has audited the source for connections to other Mozilla services.

**Comment by wchen342** [comment]
Mozilla just laid off a large portion of its team. I cannot find whether that included their push service but judging from the recent moves from Mozilla I feel using their servers is no longer a good option.

**Comment by sijanec** [comment]
Any updates? How can I self host a push notifications server for my ungoogled-chromium?

**Comment by taylorgoolsby** [comment]
Since this is a discussion now, I feel like it's okay to ask this here. So, Google and Mozilla are paying for the server fees in running push servers for their respective browsers, chrome and firefox? Anyone have any idea how much they pay for this to service all push notification usage in chrome or firefox?

If I wanted to make a browser which was open source, private, and free, but also feature complete, then I want to support push notifications, but I don't want to pay out of pocket to run my own push servers.

Is there some alternative method I could use to have push notifications in my browser without having to pay for it?

One option is to remove push servers from the push API spec. This means asking application servers to manage subscriptions and delivery of notifications. If push servers are removed, then any work that push servers had to do, like managing subscriptions and delivery, must be now handled by the application server instead. A new spec which outline how to implement this would be needed.

Otherwise, if I don't want to create a new spec to replace the push API with a new API that doesn't costs to browser owners, what else could I do?

**Comment by arjpar** [comment]
We could figure out a way how to use [autopush-rs](https://github.com/mozilla-services/autopush-rs/) by Mozilla.

I think a good idea is the one suggested by @Eloston [here](https://github.com/ungoogled-software/ungoogled-chromium/discussions/1900#discussioncomment-2539310). It can opt-in and you can specify what push server you want to use (Mozilla's or Google's).

**Reply by PF4Public** [reply]
Have you figured it out?
